graph/pull_v2.go

```diff
@@ -171,24 +179,45 @@ func (p *v2Puller) download(di *downloadInfo) {
 		ID:        stringid.TruncateID(di.img.ID),
 		Action:    "Downloading",
 	})
-	io.Copy(tmpFile, reader)
+
+	downloadComplete := make(chan error)
+	readCanceler := ioutils.NewReadCloserWithError(reader)
+
+	go func() {
+		_, err := io.Copy(tmpFile, readCanceler)
+		downloadComplete <- err
+	}()
+
+	select {
+	case err = <-downloadComplete:
+		// The download completed! We need to check the error below.
+	case <-di.cancel:
+		// We need to cancel the download. Closing the readCanceler
+		// will cause the io.Copy to fail, and we should get a non-nil
+		// error sent on the downloadComplete channel.
+		readCanceler.CloseWithError(errors.New("canceled"))
+		err = <-downloadComplete
```

Comment:
Do we really need a `ReadCloserWithError` for this code path instead of any regular `ReadCloser`? It seems to me that we could simply do:
- `Close()` -> trigger `io.Copy` to fail
- Read `<-downloadComplete` (or not?), ignore error code
- Set  `err` to `errors.New("canceled")`